system/Traits/ConditionalTrait.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Traits;
+
+trait ConditionalTrait
+{
+    /**
+     * Only runs the query when $condition evaluates to true
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     */
+    public function when($condition, callable $callback, ?callable $defaultCallback = null): self
+    {
+        if ($condition) {
+            $callback($this, $condition);
+        } elseif ($defaultCallback) {
+            $defaultCallback($this);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Only runs the query when $condition evaluates to false
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     */
```

Comment:
```suggestion
     * @param array|bool|float|int|object|resource|string|null $condition
     *
     * @return $this
     */
```

Reply:
I'll commit, but I thought we weren't adding parameter or return types to docblocks when we could add them to the method itself now?

Reply:
This is a special case. Because we cannot declare return type `$this`.

Reply:
I was just providing the suggestion for what @kenjis commented above. We don't get `static` as a return type until PHP 8 so the best we can do is use native `self` with a docblock of `$this`.